#197 Snackbar

VladislavAntonyuk opened the pull request:
### Description of Change ###

Add infrustructure for adding actions (Snackbar, other alerts)

### Bugs Fixed ###

- Fixes #197 

### API Changes ###

List all API changes here (or just put None), example:

Added: 
 
- `SnackbarArguments.shared.cs`
- `Snackar.platform.cs;`
-  `PageExtension.shared.cs`


### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [ ] Changes adhere to coding standard
- [ ] Updated documentation

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.android.cs:
```diff
@@ -0,0 +1,66 @@
+	{
+		public SnackBar(Activity subscriber)
+		{
+			MessagingCenter.Subscribe<Page, SnackbarArguments>(subscriber, PageExtension.SnackbarSignalName, OnSnackbarRequested);
```
Should we unsubscribe this event anywhere?

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.android.cs:
```diff
@@ -0,0 +1,66 @@
+
+namespace Xamarin.CommunityToolkit.Actions.Snackbar
+{
+	using global::Android.App;
```
just a question: is there any purpose why we put this using inside the namespace?

VladislavAntonyuk replied:
sorry, my default VS settings

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.android.cs:
```diff
@@ -0,0 +1,66 @@
+		}
+	}
+
+	class SnackbarCallback : AndroidSnackbar.BaseCallback
```
would be great to have an own file for each class.

VladislavAntonyuk replied:
it is private class only for android. I prefer to keep it here.

AndreiMisiukevich replied:
but seems it's an internal class

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.android.cs:
```diff
@@ -0,0 +1,66 @@
+			snackTextView.SetMaxLines(10);
+
+			if (!string.IsNullOrEmpty(arguments.ActionButtonText) && arguments.Action != null)
+			{
```
We are keen to not use braces for one like if-else body

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/SnackbarArguments.shared.cs:
```diff
@@ -0,0 +1,48 @@
+using System.ComponentModel;
+using System.Threading.Tasks;
+
+namespace Xamarin.CommunityToolkit.Actions.Snackbar
```
I am not sure about the namespace for this class.

Should this class be visible for the XCT consumers?
If not, probably we should put [EditorBrowasble] attribute over the class definition.

AndreiMisiukevich commented:
Awesome work! I am sure this will be a super popular feature in the XCT. Thank you very much for your contribution

pictos commented on XamarinCommunityToolkit/Actions/Snackbar/macOS/SnackbarAppearance.cs:
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using AppKit;
+
+namespace Xamarin.Forms.Platform.macOS.Controls.Snackbar
```
Can you please change this namespace?

pictos commented on XamarinCommunityToolkit/Actions/Snackbar/macOS/SnackbarLayout.cs:
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+
+namespace Xamarin.Forms.Platform.macOS.Controls.Snackbar
```
Can you please change this namespace?

pictos reviewed (changes requested):
Hey @VladislavAntonyuk, thank you for your contribution, migrate a control from XF to here is a hard job. Most of my requests are about naming and code style, here I'll address more general concerns.

1. I see that doesn't have the UWP implementation, any reason to don't implement it?

2. Please remember to remove the `Mac` project before merging it, we already have a [PR open](https://github.com/xamarin/XamarinCommunityToolkit/pull/242) to implement the MacOS project into the solution.

4. Can you move the Snackbar implementation into the Views folder? I see it more like control than an action.

5. To keep the implementation organized, can you move the `.platform` files to folders?

6. Can you move the Snackbar sample page, to [here](https://github.com/xamarin/XamarinCommunityToolkit/blob/main/XamarinCommunityToolkitSample/Pages/Views/AvatarViewPage.xaml)?

7. I see that this feature doesn't have tests, any reason for it?

VladislavAntonyuk commented:
@pictos, as for uwp and wpf if you open original pr, I modified xml templates and added additional space for Snackbar. I haven’t investigated to much XFC project but could you please suggest how to inject it to the page?
Also as for wpf, gtk and other platforms. I would like to keep the code even if we don’t support it. What do you think about adding net471 and netcoreapp or even net5.0 to the target framework?

pictos commented:
@VladislavAntonyuk 
> as for uwp and wpf if you open original pr, I modified xml templates and added additional space for Snackbar. I haven’t investigated to much XFC project, but could you please suggest how to inject it to the page?

I did something similar to the CameraView Android's implementation, because exists a fragment layout in `axml` so I need to configure the `csproj` to accept it. I'll do some research on UWP/WPF how this works and try to come with some idea. For now, if you can add the UWP/WPF implementation with the xml (even without any build action) would be good.
If you want to give a try, you can see my [implementation here](https://github.com/xamarin/XamarinCommunityToolkit/pull/262/files#diff-ae22bdf576b3e77bce150203e5c52331R88)

> What do you think about adding net471 and netcoreapp or even net5.0 to the target framework?

I'm can't make this decision, but I would recommend to not add net5.0 yet, the net471 should be ok and I have no idea about the netcore. @jfversluis thoughts?

pictos commented:
@VladislavAntonyuk what do you think about creating a new branch and storage the Snackbar implementation for all platforms and in the `main` branch we just have the code for the supported platforms?

VladislavAntonyuk commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.gtk.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Xamarin.CommunityToolkit.Actions.Snackbar
+{
+	public class SnackBar
```
I would like to keep it, even though we do not support it, it won't be complied in the final dll

AndreiMisiukevich commented on XamarinCommunityToolkit.UnitTests/Actions/Snackbar/Snackbar_Tests.cs:
```diff
@@ -0,0 +1,24 @@
+		public async void SnackbarCreate_PlatformNotSupportedException()
+		{
+			var page = new ContentPage();
+			await Assert.ThrowsAsync<PlatformNotSupportedException>(() => page.DisplaySnackbar("message"));
```
Hm, do we actually want to keep this unit test?

VladislavAntonyuk commented:
> Hey @VladislavAntonyuk, thank you for your contribution, migrate a control from XF to here is a hard job. Most of my requests are about naming and code style, here I'll address more general concerns.
> 
> 1. I see that doesn't have the UWP implementation, any reason to don't implement it?
> 2. Please remember to remove the `Mac` project before merging it, we already have a [PR open](https://github.com/xamarin/XamarinCommunityToolkit/pull/242) to implement the MacOS project into the solution.
> 3. Can you move the Snackbar implementation into the Views folder? I see it more like control than an action.
> 4. To keep the implementation organized, can you move the `.platform` files to folders?
> 5. Can you move the Snackbar sample page, to [here](https://github.com/xamarin/XamarinCommunityToolkit/blob/main/XamarinCommunityToolkitSample/Pages/Views/AvatarViewPage.xaml)?
> 6. I see that this feature doesn't have tests, any reason for it?

1. Still in progress. Try to find a way to imbed snackbar into the XF template
2. Done
3. Why do you think it's View? From my point of view, View is where we create control with renders. Here for example on Android I just call SnackBar.Make. The same for Tizen. we create dialog
4. I have only folders for iOS and macOS, so keeping snackbar.platform.cs in the same folder I think looks better.
5. Done
6. Added UT for page extension.

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.macos.cs:
```diff
@@ -0,0 +1,33 @@
+﻿using System.Threading.Tasks;
```
Can we leave only one file for both macos and iOS?

Snackbar.macos.ios.cs

VladislavAntonyuk replied:
they are similar, but they used different Snackbar: IOSSnackbar and MacOSSnackBar

AndreiMisiukevich replied:
Yeah, I see. But I believe we can resolve it by 

```
#if __IOS__
// iOS-specific code
#elif
// macOS
#endif
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.tizen.cs:
```diff
@@ -0,0 +1,44 @@
+
+			if (!string.IsNullOrEmpty(arguments.ActionButtonText) && arguments.Action != null)
+			{
+				var ok = new EButton(_snackbarDialog) {Text = arguments.ActionButtonText};
```
```suggestion
				var ok = new EButton(_snackbarDialog) { Text = arguments.ActionButtonText };
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/SnackbarArguments.shared.cs:
```diff
@@ -0,0 +1,46 @@
+using System.ComponentModel;
+using System.Threading.Tasks;
+
+namespace Xamarin.CommunityToolkit.Actions.Snackbar
```
@jfversluis where do you think such stuff should live? Could you please hint

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/iOS/Extensions/UIViewExtensions.cs:
```diff
@@ -0,0 +1,77 @@
+﻿using UIKit;
```
I guess the file name should be UIViewExtensions.ios.cs, right?

VladislavAntonyuk replied:
I added the whole folder to Compile on iOS. Do you think it would be better to rename all files?

AndreiMisiukevich replied:
Ah, yeah, you are right!
No, I think that's fine now

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/iOS/IOSSnackBar.cs:
```diff
@@ -0,0 +1,108 @@
+﻿using System;
```
Same here IOSSnackBar.ios.cs

AndreiMisiukevich commented on XamarinCommunityToolkitSample/Pages/Views/ActionsPage.xaml:
```diff
@@ -0,0 +1,11 @@
+                xmlns:pages="clr-namespace:Xamarin.CommunityToolkit.Sample.Pages"
+                x:Class="Xamarin.CommunityToolkit.Sample.Pages.Views.ActionsPage">
+    <StackLayout>
+        <Button Clicked="Button1_Clicked" Text="Show snackbar with action button"></Button>
```
Please add these texts to Resx file

AndreiMisiukevich commented on XamarinCommunityToolkitSample/Pages/Views/ActionsPage.xaml.cs:
```diff
@@ -0,0 +1,45 @@
+		async void Button2_Clicked(object sender, EventArgs args)
+		{
+			var result = await this.DisplaySnackbar(GenerateLongText(5));
+			StatusText.Text = result ? "Snackbar is closed by user" : "Snackbar is closed by timeout";
```
All texts to Resx please

AndreiMisiukevich reviewed (commented):
That's amazing work! Well done

Just a question. Should all iOS/macOS files has platform inside the file FileName**.ios.**cs

jfversluis reviewed (changes requested):
I see a commit that says to remove WPF and GTK, But it still see them in the changes? Also please remove the `libs` folder with all the GTK binaries

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.gtk.cs:
```diff
@@ -0,0 +1,74 @@
+		}
+
+		static Widget GetTopWindowContainer(Widget platformRenderer)
+		{
+			return (platformRenderer.Toplevel as Window).Child;
+		}
```
```suggestion
			=> return (platformRenderer.Toplevel as Window).Child;
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.shared.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Actions.Snackbar
```
@jfversluis please take a look at the namespaces. I am not sure about this one)

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/Snackbar.uwp.cs:
```diff
@@ -0,0 +1,87 @@
+		{
+			public ExtendedPageControl(PageControl pageControl)
+			{
+
```
```suggestion
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Actions/Snackbar/iOS/Extensions/UIViewExtensions.ios.cs:
```diff
@@ -0,0 +1,77 @@
+		public static NSLayoutYAxisAnchor SafeBottomAnchor(this UIView view)
+		{
+			return UIDevice.CurrentDevice.CheckSystemVersion(11, 0)
+				? view.SafeAreaLayoutGuide.BottomAnchor
```
Would you mind to update this file to follow the code-style with lambda-method-body for one line methods.

```
=> UIDevice.CurrentDevice.CheckSystemVersion(11, 0)
	? view.SafeAreaLayoutGuide.BottomAnchor
	: view.BottomAnchor;
```

Also, it makes sense to make a separate property **IsIos11OrHigher** and use it

AndreiMisiukevich reviewed (commented):
My picky notes can be left without attention, I believe.
This is great work, congrats!

Just wonder whether we want to add WPF & **GTK** code with this PR
And what namespace should we use for this feature

AndreiMisiukevich commented:
@VladislavAntonyuk could you please resolve merge conflicts?

AndreiMisiukevich commented on XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <TargetFrameworks>netstandard1.0;netstandard2.0;netstandard2.1;Xamarin.iOS10;MonoAndroid90;MonoAndroid10.0;Xamarin.TVOS10;Xamarin.WatchOS10;Xamarin.Mac20;tizen40</TargetFrameworks>
-    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);uap10.0.16299</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">$(TargetFrameworks);uap10.0.17763;net472</TargetFrameworks>
```
**net472** the reason why the build pipeline failed.

@jfversluis do we want to support GTK/WPF? If yes, then we should come up with the solution regarding to **ValueTask**. (The build failed because ValueTask is not available for net472)

bartlannoeye replied:
Can be added with https://www.nuget.org/packages/System.Threading.Tasks.Extensions/

AndreiMisiukevich replied:
Then we should add this nuget for net472

VladislavAntonyuk replied:
give me a minute, I am about to fix it

VladislavAntonyuk replied:
I do not add GTK (net471) target framework. the same for WPF (net5-windows or netcoreapp3.1)

pictos replied:
I saw that you changed the UWP version, any reason for that?

VladislavAntonyuk replied:
XF recommends version 17763

pictos commented on XamarinCommunityToolkit/Actions/Snackbar/iOS/SnackbarLayout.ios.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Xamarin.CommunityToolkit.Actions.Snackbar.iOS
+{
+	class SnackbarLayout
```
This class looks exactly the like [this one ](https://github.com/xamarin/XamarinCommunityToolkit/pull/261/files#diff-7f717804b106b22365cfc3e752af5545R5). Can you keep just one and rename this file to
 SnackbarLayout.ios.macos.cs

pictos commented on XamarinCommunityToolkit/Actions/Snackbar/macOS/SnackbarViews/ActionMessageSnackbarView.macos.cs:
```diff
@@ -0,0 +1,69 @@
+			AddSubview(ActionButton);
+		}
+
+		async void DismissButton_TouchUpInside(object sender, EventArgs e)
```
can you please remove the `_`

pictos commented on XamarinCommunityToolkit/Xamarin.CommunityToolkit.csproj:
```diff
@@ -132,5 +135,18 @@
+  <ItemGroup Condition=" $(TargetFramework.StartsWith('net472')) ">
+    <Compile Include="**\*.wpf.cs" />
+    <Compile Include="**\*.wpf.*.cs" />
+    <Compile Remove="**\Snackbar.shared.cs" />
```
do we need this line?

VladislavAntonyuk replied:
We need to remove shared file from all platforms

VladislavAntonyuk replied:
We need to remove shared file from all platforms

pictos commented on XamarinCommunityToolkitSample.UWP/Xamarin.CommunityToolkit.Sample.UWP.csproj:
```diff
@@ -12,7 +12,7 @@
     <TargetPlatformIdentifier>UAP</TargetPlatformIdentifier>
     <TargetPlatformVersion>10.0.18362.0</TargetPlatformVersion>
-    <TargetPlatformMinVersion>10.0.16299.0</TargetPlatformMinVersion>
+    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
```
Any reason to increase the minVersion?

VladislavAntonyuk replied:
To avoid warning during the build (XF recommends min version 17763)

pictos reviewed (changes requested):
@VladislavAntonyuk were you able to add the XML templates to WPF? I'm sorry I did not have time to look at it since I'm moving to my hometown. Probably I'll have time next weekend.

jfversluis commented:
Awesome work! Sorry for the late reply @VladislavAntonyuk, I had some time off.

I see two things from @pictos would you be able to solve those? I think we are ready to merge then! :)

Thank you for all your efforts!
